TrailBlaze/Networking/Interceptor.m

```diff
@@ -0,0 +1,125 @@
+
+        MKMapItem *destinationItem =  [[MKMapItem alloc] initWithPlacemark:destinationPlacemark];
+
+        MKDirectionsRequest *pathInterceptorRequest = [[MKDirectionsRequest alloc] init];
+        [pathInterceptorRequest setSource:startInterceptorItem];
+        [pathInterceptorRequest setDestination:destinationItem];
+        [pathInterceptorRequest setTransportType:MKDirectionsTransportTypeWalking];
+        [pathInterceptorRequest setRequestsAlternateRoutes:NO];
+        
+        MKDirectionsRequest *pathRunnerRequest = [[MKDirectionsRequest alloc] init];
+        [pathRunnerRequest setSource:startRunnerItem];
+        [pathRunnerRequest setDestination:destinationItem];
+        [pathRunnerRequest setTransportType:MKDirectionsTransportTypeWalking];
+        [pathRunnerRequest setRequestsAlternateRoutes:NO];
+        
+        MKDirections *pathInterceptor = [[MKDirections alloc] initWithRequest:pathInterceptorRequest];
+        MKDirections *pathRunner = [[MKDirections alloc] initWithRequest:pathRunnerRequest];
+        [pathInterceptor calculateETAWithCompletionHandler:^(MKETAResponse * _Nullable interceptorResponse, NSError * _Nullable error1) {
+            if (interceptorResponse) {
+                [pathRunner calculateETAWithCompletionHandler:^(MKETAResponse * _Nullable runnerResponse, NSError * _Nullable error2) {
+                    if (runnerResponse) {
+                        [etasDifferences addObject:[NSNumber numberWithDouble:fabs(interceptorResponse.expectedTravelTime - runnerResponse.expectedTravelTime)]];
+                        [etaPointPairs setValue:runnerResponse.destination forKey:[NSString stringWithFormat:@"%.f", fabs(interceptorResponse.expectedTravelTime - runnerResponse.expectedTravelTime)]];
+                    } else {
+                        NSLog(@"%@", error2.localizedDescription);
+                    }
+                    dispatch_async(serialQueue, ^{
+                        dispatch_group_leave(group);
+                    });
+                }];
+            } else {
+                NSLog(@"%@", error1.localizedDescription);
+            }
+        }];
+    }
+    dispatch_group_notify(group, dispatch_get_main_queue(), ^{
+        NSLog(@"everything is done!!");
+        NSLog(@"%@", etasDifferences);
+        NSLog(@"%@", [self sortInAscendingOrder:etasDifferences]);
+//                    NSLog(@"%@",
+        NSLog(@"%@", [NSString stringWithFormat:@"%@", [[self sortInAscendingOrder:etasDifferences] firstObject]]);
+        NSNumber *bestDifference = [[self sortInAscendingOrder:etasDifferences] firstObject];
+        if ([bestDifference floatValue] <= etaDifferenceThreshold) {
+        completion([etaPointPairs objectForKey:[NSString stringWithFormat:@"%@", bestDifference]], nil);
+        } else {
+            completion(nil, nil);
```

Comment:
let's return some `Error` here. a completion block that returns something _or_ an error usually implies that you'll get 1 or the other and not neither (or both).

Reply:
There is a case where the function doesn't error out but the lowest eta difference happens to be higher than the etaDifferenceThreshold, for this case an intercept will not be possible but no error is thrown. I therefore have it set that in this case the completion returns nil and the HomeViewController knows to prompts the user of the lack of an intercept route when the completion returns nil.

Reply:
> There is a case where the function doesn't error out

the internal `calculateETA` operations don't fail but i would consider `getBestETAPoint` returning a `nil` best point to be a failure. in this case client's should get a nonnull error back. you can create one directly like:

```
[[NSError alloc] initWithDomain:@"Interceptor":
                                   code:0
                               userInfo:@{NSLocalizedDescriptionKey : @"No intercept route below threshold found"}];
```

there's a concept of `Result` type in swift that illustrates this nicely. basically a completion returns a "result", which is a success or a failure. a success gives you a nonnull success object, a failure gives you a nonnull Error value. https://www.hackingwithswift.com/articles/161/how-to-use-result-in-swift